### PR TITLE
FEATURE: add ProfitStatsTracker

### PR DESCRIPTION
**Pull request by andycheng123**

### CODE REVIEW

File: pkg/report/profit_tracker.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	*p.currentProfitStats = types.NewProfitStats(p.market)
+	p.profitStatsSlice = append(p.profitStatsSlice, *p.currentProfitStats)
+	// Truncate
+	if len(p.profitStatsSlice) > p.Window {
+		p.profitStatsSlice = p.profitStatsSlice[len(p.profitStatsSlice)-p.Window:]
+	}
+}
+
+func (p *ProfitTracker) AddProfit(profit types.Profit) {
+	(*p.currentProfitStats).AddProfit(profit)
```
Comment (c9s) [inline pkg/report/profit_tracker.go:62]: could be just `p.currentProfitStats.AddProfit(profit)`

File: config/supertrend.yaml [hunk 2]
```diff
@@ -112,3 +112,11 @@ exchangeStrategies:
           # If true, looking for lower lows in long position and higher highs in short position. If false, looking for
           # higher highs in long position and lower lows in short position
           oppositeDirectionAsPosition: false
+
+    profitTracker:
+      Interval: 1d
```
Comment (c9s) [inline config/supertrend.yaml:117]: should be in lower case?

File: config/supertrend.yaml [hunk 3]
```diff
@@ -112,3 +112,11 @@ exchangeStrategies:
...
+      Window: 30
```
Comment (c9s) [inline config/supertrend.yaml:118]: here too

File: pkg/bbgo/order_executor_general.go [hunk 4]
```diff
@@ -162,6 +163,27 @@ func (e *GeneralOrderExecutor) BindProfitStats(profitStats *types.ProfitStats) {
 	})
 }
 
+func (e *GeneralOrderExecutor) BindProfitTracker(profitTracker *report.ProfitTracker) {
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:166]: hmm move this to the strategy? I think it's bit too much for the order executor

File: pkg/bbgo/order_executor_general.go [hunk 5]
```diff
@@ -162,6 +163,27 @@ func (e *GeneralOrderExecutor) BindProfitStats(profitStats *types.ProfitStats) {
...
+	e.session.Subscribe(types.KLineChannel, profitTracker.Market.Symbol, types.SubscribeOptions{Interval: profitTracker.Interval})
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:167]: note the "Subscribe" call must be done before we connect the websocket server

File: pkg/report/profit_report.go [hunk 6]
```diff
@@ -12,134 +11,104 @@ import (
...
-	NumberOfInterval int `json:"NumberOfInterval"`
+	// ShortTermProfitWindow The window to sum up the short-term profit
+	ShortTermProfitWindow int `json:"shortTermProfitWindow"`
 
 	// TsvReportPath The path to output report to
 	TsvReportPath string `json:"tsvReportPath"`
 
-	// AccumulatedDailyProfitWindow The window to sum up the daily profit, in days
-	AccumulatedDailyProfitWindow int `json:"accumulatedDailyProfitWindow"`
+	symbol string
```
Comment (c9s) [inline pkg/report/profit_report.go:23]: why apply private field here?

File: pkg/report/profit_report.go [hunk 7]
```diff
@@ -12,134 +11,104 @@ import (
...
-	accumulatedProfit         fixedpoint.Value
-	accumulatedProfitPerDay   floats.Slice
-	previousAccumulatedProfit fixedpoint.Value
+	accumulatedProfit            fixedpoint.Value
+	accumulatedProfitPerInterval floats.Slice
 
 	// Accumulated profit MA
-	accumulatedProfitMA       *indicator.SMA
-	accumulatedProfitMAPerDay floats.Slice
+	profitMA            *indicator.SMA
```
Comment (c9s) [inline pkg/report/profit_report.go:32]: let's use the v2 indicator?

File: pkg/report/profit_tracker.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
+package report
+
+import (
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+type ProfitTracker struct {
```
Comment (c9s) [inline pkg/report/profit_tracker.go:7]: name this ProfitStatsTracker?

File: pkg/types/profit.go [hunk 9]
```diff
@@ -164,6 +164,9 @@ type ProfitStats struct {
 	TodayGrossProfit fixedpoint.Value `json:"todayGrossProfit,omitempty"`
 	TodayGrossLoss   fixedpoint.Value `json:"todayGrossLoss,omitempty"`
 	TodaySince       int64            `json:"todaySince,omitempty"`
+
+	startTime time.Time
```
Comment (c9s) [inline pkg/types/profit.go:168]: these could be public fields

File: pkg/types/profit.go [hunk 10]
```diff
@@ -182,6 +185,8 @@ func NewProfitStats(market Market) *ProfitStats {
 		TodayGrossProfit:       fixedpoint.Zero,
 		TodayGrossLoss:         fixedpoint.Zero,
 		TodaySince:             0,
+		startTime:              time.Now().UTC(),
+		endTime:                time.Now().UTC(),
```
Comment (c9s) [inline pkg/types/profit.go:189]: could be errored if you're running backtesting.

File: pkg/bbgo/order_executor_general.go (hunk 5 above)
Comment (andycheng123) [reply pkg/bbgo/order_executor_general.go:167]: Do you mean put those subscribe call in Strategy.Subscribe() ? Is it executed before connecting to the websocket server?

File: pkg/types/profit.go (hunk 10 above)
Comment (andycheng123) [reply pkg/types/profit.go:189]: ah! so how do i get the "now" in backtesting?

File: pkg/bbgo/order_executor_general.go (hunk 5 above)
Comment (c9s) [reply pkg/bbgo/order_executor_general.go:167]: yes

File: pkg/types/profit.go (hunk 10 above)
Comment (c9s) [reply pkg/types/profit.go:189]: interesting, I guess we have the time in the bbgo.Environment, if not, we will need to add it LOL

File: config/supertrend.yaml (hunk 2 above)
Comment (andycheng123) [reply config/supertrend.yaml]: done

File: config/supertrend.yaml (hunk 3 above)
Comment (andycheng123) [reply config/supertrend.yaml]: done

File: pkg/bbgo/order_executor_general.go (hunk 4 above)
Comment (andycheng123) [reply pkg/bbgo/order_executor_general.go:166]: done

File: pkg/bbgo/order_executor_general.go (hunk 5 above)
Comment (andycheng123) [reply pkg/bbgo/order_executor_general.go:167]: done

File: pkg/report/profit_report.go (hunk 6 above)
Comment (andycheng123) [reply pkg/report/profit_report.go:23]: The symbol of the profit report is determined by the strategy

File: pkg/report/profit_tracker.go (hunk 8 above)
Comment (andycheng123) [reply pkg/report/profit_tracker.go:7]: done

File: pkg/types/profit.go (hunk 10 above)
Comment (andycheng123) [reply pkg/types/profit.go:189]: nope, we don't have "now" in bbgo.Environment. we have StartTime() only. I just commented out those lines related to time. will add them back once we have "now" in bbgo.Environment

**Comment by andycheng123** [comment]
@c9s there's a build error i don't understand:
Error: pkg/report/profit_stats_tracker.go:51:87: undefined: bbgo.TradeCollector
would you kindly help me out?

File: pkg/report/profit_report.go [hunk 11]
```diff
@@ -150,46 +119,34 @@ func (r *AccumulatedProfitReport) Output(symbol string) {
...
 				fmt.Sprintf("%d", i+1),
-				symbol, accumulatedProfitStr,
-				accumulatedProfitMAStr,
-				intervalAccumulatedProfitStr,
-				accumulatedFee,
-				accumulatedNetProfit,
-				winRatio, profitFactor,
-				tradesStr,
+				r.symbol,
+				fmt.Sprintf("%f", r.accumulatedProfitPerInterval.Last(i)),
```
Comment (c9s) [inline pkg/report/profit_report.go:140]: you can use strconv.FormatFloat64

File: pkg/report/profit_report.go [hunk 12]
```diff
@@ -150,46 +119,34 @@ func (r *AccumulatedProfitReport) Output(symbol string) {
...
+				fmt.Sprintf("%f", r.accumulatedProfitPerInterval.Last(i)-r.accumulatedProfitPerInterval.Last(i+r.ShortTermProfitWindow)),
+				fmt.Sprintf("%f", r.accumulatedFeePerInterval.Last(i)),
+				fmt.Sprintf("%f", r.winRatioPerInterval.Last(i)),
+				fmt.Sprintf("%f", r.profitFactorPerInterval.Last(i)),
+				fmt.Sprintf("%f", r.accumulatedTradesPerInterval.Last(i)),
 			}
-			for j := 0; j < len(r.extraValues); j++ {
-				values = append(values, r.extraValues[j][1])
+			for j := 0; j < len(r.strategyParameters); j++ {
+				values = append(values, r.strategyParameters[j][1])
```
Comment (c9s) [inline pkg/report/profit_report.go:149]: Actually you should use the CsvHeader, CsvRecords interface to output this kind of data

File: pkg/report/profit_stats_tracker.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+	tradeStats *types.TradeStats
+}
+
+func (p *ProfitStatsTracker) Subscribe(session *bbgo.ExchangeSession, symbol string) {
+	session.Subscribe(types.KLineChannel, symbol, types.SubscribeOptions{Interval: p.Interval})
+}
+
+// InitOld is for backward capability. ps is the ProfitStats of the strategy, Market is the strategy Market
+func (p *ProfitStatsTracker) InitOld(market types.Market, ps **types.ProfitStats, ts *types.TradeStats) {
```
Comment (c9s) [inline pkg/report/profit_stats_tracker.go:29]: How about InitLegacy?

File: pkg/report/profit_report.go (hunk 11 above)
Comment (andycheng123) [reply pkg/report/profit_report.go:140]: done

File: pkg/report/profit_report.go [hunk 14]
```diff
@@ -150,46 +119,34 @@ func (r *AccumulatedProfitReport) Output(symbol string) {
...
-			intervalAccumulatedProfitStr := fmt.Sprintf("%f", intervalAccumulatedProfit)
-			accumulatedFee := fmt.Sprintf("%f", r.accumulatedFeePerDay.Index(r.IntervalWindow*i))
-			accumulatedNetProfit := fmt.Sprintf("%f", accumulatedProfit-r.accumulatedFeePerDay.Index(r.IntervalWindow*i))
-			winRatio := fmt.Sprintf("%f", r.winRatioPerDay.Index(r.IntervalWindow*i))
-			profitFactor := fmt.Sprintf("%f", r.profitFactorPerDay.Index(r.IntervalWindow*i))
-			trades := r.dailyTrades.Tail(60+r.IntervalWindow*i).Sum() - r.dailyTrades.Tail(r.IntervalWindow*i).Sum()
-			tradesStr := fmt.Sprintf("%f", trades)
+		for i := 0; i <= r.Window-1; i++ {
 			values := []string{
 				fmt.Sprintf("%d", i+1),
```
[the remark this answers is not in the document]
Comment (andycheng123) [reply pkg/report/profit_report.go:138]: done

File: pkg/report/profit_stats_tracker.go (hunk 13 above)
Comment (andycheng123) [reply pkg/report/profit_stats_tracker.go]: done

File: pkg/report/profit_report.go (hunk 12 above)
Comment (andycheng123) [reply pkg/report/profit_report.go:150]: done

**Review by c9s — APPROVED** [review]

**Pull request merged**